File: day15/day15.py
#!/usr/bin/env python

# python script to answer questions from http://adventofcode.com/2017/day/15

# lazily evaluated zip and infinite generator limiter
from itertools import izip, islice
from __future__ import print_function

# puzzle inputs
inputA = 679
inputB = 771

# test inputs
testA = 65
testB = 8921

def gen(start, factor):
    '''(int, int) => generator of int
    Yield sequence of generator defined by `factor` and starting with `start`
    '''
    val = start
    while True:
        val = (val * factor) % 2147483647
        yield val

def judge_equal(tpl):
    '''(tuple of (int, int)) => bool
    Return True if the last 16 bits of `tpl` entries are equal
    '''
    a, b = tpl
    return (a % 2**16) == (b % 2**16)


def start_generators(startA, startB):
    '''(int, int) => generator of (int, int)
    Start both generators and return an interable of values from each one (a, b)
    '''
    genA = gen(startA, 16807)
    genB = gen(startB, 48271)
    return (x for x in izip(genA, genB))

# test generators
test_gens = start_generators(testA, testB)

# get first 5 values of each
for _ in range(5):
    print(next(test_gens))
    
# event better, use and explicit iterator
for x in islice(test_gens, 5):
    print(x)


# restart test generators and get count of matches, expected 588
test_gens = start_generators(testA, testB)
sum((judge_equal(p) for p in islice(test_gens, 40000000)))

####################
##  First puzzle  ##
####################

gens = start_generators(inputA, inputB)
first_solution = sum((judge_equal(p) for p in islice(gens, 40000000)))
print('The answer to the first puzzle is {}'.format(first_solution))

#####################
##  Second puzzle  ##
#####################

# we need to redefine the procedure for starting the generators
def start_generators_filtered(startA, startB, critA, critB):
    '''(int, int, int, int) => generator of (int, int)
    Start both generators, apply the  and return an interable of values from each one (a, b)
    '''
    genA = gen(startA, 16807)
    genA_filter = (x for x in genA if x % critA == 0)
    
    genB = gen(startB, 48271)
    genB_filter = (x for x in genB if x % critB == 0)
    
    return (x for x in izip(genA_filter, genB_filter))


# start generators with filtering
gens_filter_test = start_generators_filtered(testA, testB, 4, 8)

# try the first few entries
for x in islice(gens_filter_test, 5):
    print(x)

# start proper generators
gens_filter = start_generators_filtered(inputA, inputB, 4, 8)
second_solution = sum((judge_equal(p) for p in islice(gens_filter, 5000000)))
print('The answer to the second puzzle is {}'.format(second_solution))

File: day15/day15.scala
#!/usr/bin/env scala

// setup inputs
val inputA = 679
val inputB = 771

val testA = 65
val testB = 8921

// constructor of iteration step for a generator
def get_next(factor: Int)(x: BigInt): BigInt = (x * factor) % 2147483647

// costruct iterator for both generators
def nextA(x: BigInt):BigInt = get_next(16807)(x)
def nextB(x: BigInt):BigInt = get_next(48271)(x)

// test iterator
nextA(1092455)       // expected 1181022009
nextB(1233683848)    // expected 1431495498

// stream constructor
def gen(next: BigInt => BigInt, start:BigInt): Stream[BigInt] = {
    val first_value = next(start)
    first_value #:: gen(next, first_value)
}

// comparison function, return one if there's a match
val last16bits = math.pow(2, 16).toInt
def judge_equal(tpl:(BigInt, BigInt)):Int = if ((tpl._1 % last16bits) == (tpl._2 % last16bits)) 1 else 0

// build generators
val genAtest = gen(nextA, testA)
val genBtest = gen(nextB, testB)

// count equals over 40m pairs, expected 588
// crashes with OOM error 
val num_pairs = 4e7.toInt
(genAtest zip genBtest) take(num_pairs) map(judge_equal) reduceLeft (_ + _)


/*******************
**  First puzzle  **
*******************/

// build generators
val genA = gen(nextA, inputA)
val genB = gen(nextB, inputB)

val first_solution = ((genA zip genB)
                        .take(num_pairs)
                        .map(judge_equal)
                        .sum)
             
print("The soluton to the first puzzle is " + first_solution.toString)


/********************
**  Second puzzle  **
********************/

// build test generators with filter
val genAtest = gen(nextA, testA) filter (_ % 4 == 0)
val genBtest = gen(nextB, testB) filter (_ % 8 == 0)
(genAtest zip genBtest) take 5 foreach println

// build generators
val genAfilter = gen(nextA, inputA) filter (_ % 4 == 0)
val genBfilter = gen(nextB, inputB) filter (_ % 8 == 0)

// this one also crashes with OOM error
val num_pairs = 5e6.toInt
val second_solution = ((genAfilter zip genBfilter)
                        .take(num_pairs)
                        .map(judge_equal)
                        .sum)
                        
print("The soluton to the second puzzle is " + second_solution.toString)


